source/fuzz/transformation_replace_id_with_synonym.cpp

```diff
@@ -0,0 +1,150 @@
+          context, use_instruction,
+          message_.id_use_descriptor().in_operand_index(),
+          message_.data_descriptor())) {
+    return false;
+  }
+
+  assert(message_.fresh_id_for_temporary() == 0);
+  assert(message_.data_descriptor().index().empty());
+
+  return true;
+}
+
+void TransformationReplaceIdWithSynonym::Apply(
+    spvtools::opt::IRContext* context,
+    spvtools::fuzz::FactManager* /*unused*/) const {
+  assert(message_.data_descriptor().index().empty());
+  auto instruction_to_change =
+      transformation::FindInstruction(message_.id_use_descriptor(), context);
+  instruction_to_change->SetInOperand(
+      message_.id_use_descriptor().in_operand_index(),
+      {message_.data_descriptor().object()});
+  context->InvalidateAnalysesExceptFor(opt::IRContext::Analysis::kAnalysisNone);
+}
+
+protobufs::Transformation TransformationReplaceIdWithSynonym::ToMessage()
+    const {
+  protobufs::Transformation result;
+  *result.mutable_replace_id_with_synonym() = message_;
+  return result;
+}
+
+bool TransformationReplaceIdWithSynonym::ReplacingUseWithSynonymIsOk(
+    opt::IRContext* context, opt::Instruction* use_instruction,
+    uint32_t use_in_operand_index, const protobufs::DataDescriptor& synonym) {
+  auto defining_instruction =
+      context->get_def_use_mgr()->GetDef(synonym.object());
+
+  if (use_instruction == defining_instruction) {
+    // If we have an instruction:
+    //   %a = OpCopyObject %t %b
+    // then we know %a and %b are synonymous, but we do *not* want to turn
+    // this into:
+    //   %a = OpCopyObject %t %a
+    // We require this special case because an instruction dominates itself.
+    return false;
+  }
+
+  if (use_instruction->opcode() == SpvOpAccessChain &&
+      use_in_operand_index > 0) {
+    // This is an access chain index.  If the object being accessed has
+    // pointer-to-struct type then we cannot replace the use with a synonym, as
+    // the use needs to be an OpConstant.
+    auto object_being_accessed = context->get_def_use_mgr()->GetDef(
+        use_instruction->GetSingleWordInOperand(0));
+    auto pointer_type =
+        context->get_type_mgr()->GetType(object_being_accessed->type_id());
+    assert(pointer_type->AsPointer());
+    if (pointer_type->AsPointer()->pointee_type()->AsStruct()) {
+      return false;
+    }
+  }
+
+  auto dominator_analysis = context->GetDominatorAnalysis(
+      context->get_instr_block(use_instruction)->GetParent());
+  if (use_instruction->opcode() == SpvOpPhi) {
+    auto parent_block =
+        use_instruction->GetSingleWordInOperand(use_in_operand_index + 1);
+    if (!dominator_analysis->Dominates(
+            context->get_instr_block(defining_instruction)->id(),
+            parent_block)) {
+      return false;
+    }
+  } else if (!dominator_analysis->Dominates(defining_instruction,
+                                            use_instruction)) {
+    return false;
+  }
+  return true;
```

Comment:
The last case is pretty clear and I know redundant comments are silly, but for consistency, perhaps you could add a comment explaining what you are checking in the last case.

Reply:
Agreed - done.